### PR TITLE
Build out a JSON API-structured API for libraries and locations 

### PR DESCRIPTION
**Pull request by cbeer**
Library info:

`GET /library/green.json`
(or, e.g. `GET /library/green.json?from=2019-08-01&to=2019-08-07`)

```json
{

   "data" : {
      "type" : "libraries",
      "links" : {
         "self" : "http://127.0.0.1:3000/libraries/green"
      },
      "attributes" : {
         "name" : "Cecil H. Green Library"
      },
      "id" : "green",
      "relationships" : {
         "locations" : {
            "data" : [
               {
                  "type" : "locations",
                  "id" : "green/information-center"
               },
               {
                  "id" : "green/library-circulation",
                  "type" : "locations"
               },
               {
                  "id" : "green/media-microtext-center",
                  "type" : "locations"
               },
               {
                  "type" : "locations",
                  "id" : "green/privileges-desk"
               },
               {
                  "id" : "green/ssds-walk-in-consulting",
                  "type" : "locations"
               },
               {
                  "type" : "locations",
                  "id" : "green/velma-denning-room"
               }
            ],
            "links" : {
               "related" : "http://127.0.0.1:3000/libraries/green/locations"
            }
         }
      }
   },
   "included" : [
      {
         "attributes" : {
            "hours" : [
               {
                  "closed" : true,
                  "weekday" : "Sunday",
                  "day" : "2019-08-04",
                  "type" : "Summer quarter"
               },
               {
                  "weekday" : "Monday",
                  "opens_at" : "2019-08-05T10:00:00.000-07:00",
                  "closed" : false,
                  "type" : "Summer quarter",
                  "closes_at" : "2019-08-05T17:00:00.000-07:00",
                  "day" : "2019-08-05"
               },
               {
                  "closed" : false,
                  "opens_at" : "2019-08-06T10:00:00.000-07:00",
                  "weekday" : "Tuesday",
                  "day" : "2019-08-06",
                  "type" : "Summer quarter",
                  "closes_at" : "2019-08-06T17:00:00.000-07:00"
               },
               {
                  "closed" : false,
                  "weekday" : "Wednesday",
                  "opens_at" : "2019-08-07T10:00:00.000-07:00",
                  "day" : "2019-08-07",
                  "type" : "Summer quarter",
                  "closes_at" : "2019-08-07T17:00:00.000-07:00"
               },
               {
                  "day" : "2019-08-08",
                  "type" : "Summer quarter",
                  "closes_at" : "2019-08-08T17:00:00.000-07:00",
                  "closed" : false,
                  "opens_at" : "2019-08-08T10:00:00.000-07:00",
                  "weekday" : "Thursday"
               },
               {
                  "opens_at" : "2019-08-09T10:00:00.000-07:00",
                  "weekday" : "Friday",
                  "closed" : false,
                  "type" : "Summer quarter",
                  "closes_at" : "2019-08-09T17:00:00.000-07:00",
                  "day" : "2019-08-09"
               },
               {
                  "type" : "Summer quarter",
                  "notes" : "End Quarter Period",
                  "day" : "2019-08-10",
                  "weekday" : "Saturday",
                  "closed" : true
               }
            ],
            "name" : "Information Center"
         },
         "links" : {
            "self" : "http://127.0.0.1:3000/libraries/green/locations/information-center"
         },
         "type" : "locations",
         "relationships" : {
            "library" : {
               "data" : {
                  "type" : "libraries",
                  "id" : "green"
               },
               "links" : {
                  "related" : "http://127.0.0.1:3000/libraries/green"
               }
            }
         },
         "id" : "green/information-center"
      },
      {
         "id" : "green/library-circulation",
         "relationships" : {
            "library" : {
               "data" : {
                  "id" : "green",
                  "type" : "libraries"
               },
               "links" : {
                  "related" : "http://127.0.0.1:3000/libraries/green"
               }
            }
         },
         "attributes" : {
            "hours" : [
               {
                  "type" : "Summer quarter",
                  "closes_at" : "2019-08-04T20:00:00.000-07:00",
                  "day" : "2019-08-04",
                  "opens_at" : "2019-08-04T12:00:00.000-07:00",
                  "weekday" : "Sunday",
                  "closed" : false
               },
               {
                  "weekday" : "Monday",
                  "opens_at" : "2019-08-05T08:00:00.000-07:00",
                  "closed" : false,
                  "type" : "Summer quarter",
                  "closes_at" : "2019-08-05T21:00:00.000-07:00",
                  "day" : "2019-08-05"
               },
               {
                  "opens_at" : "2019-08-06T08:00:00.000-07:00",
                  "weekday" : "Tuesday",
                  "closed" : false,
                  "type" : "Summer quarter",
                  "closes_at" : "2019-08-06T21:00:00.000-07:00",
                  "day" : "2019-08-06"
               },
               {
                  "day" : "2019-08-07",
                  "type" : "Summer quarter",
                  "closes_at" : "2019-08-07T21:00:00.000-07:00",
                  "closed" : false,
                  "opens_at" : "2019-08-07T08:00:00.000-07:00",
                  "weekday" : "Wednesday"
               },
               {
                  "opens_at" : "2019-08-08T08:00:00.000-07:00",
                  "weekday" : "Thursday",
                  "closed" : false,
                  "type" : "Summer quarter",
                  "closes_at" : "2019-08-08T21:00:00.000-07:00",
                  "day" : "2019-08-08"
               },
               {
                  "closes_at" : "2019-08-09T18:00:00.000-07:00",
                  "type" : "Summer quarter",
                  "day" : "2019-08-09",
                  "opens_at" : "2019-08-09T08:00:00.000-07:00",
                  "weekday" : "Friday",
                  "closed" : false
               },
               {
                  "day" : "2019-08-10",
                  "notes" : "End Quarter Period",
                  "type" : "Summer quarter",
                  "closes_at" : "2019-08-10T18:00:00.000-07:00",
                  "closed" : false,
                  "weekday" : "Saturday",
                  "opens_at" : "2019-08-10T10:00:00.000-07:00"
               }
            ],
            "name" : "Library & circulation"
         },
         "type" : "locations",
         "links" : {
            "self" : "http://127.0.0.1:3000/libraries/green/locations/library-circulation"
         }
      },
      {
         "relationships" : {
            "library" : {
               "links" : {
                  "related" : "http://127.0.0.1:3000/libraries/green"
               },
               "data" : {
                  "type" : "libraries",
                  "id" : "green"
               }
            }
         },
         "id" : "green/media-microtext-center",
         "attributes" : {
            "name" : "Media & Microtext Center",
            "hours" : [
               {
                  "closes_at" : "2019-08-04T20:00:00.000-07:00",
                  "type" : "Summer quarter",
                  "day" : "2019-08-04",
                  "opens_at" : "2019-08-04T12:00:00.000-07:00",
                  "weekday" : "Sunday",
                  "closed" : false
               },
               {
                  "closed" : false,
                  "opens_at" : "2019-08-05T08:00:00.000-07:00",
                  "weekday" : "Monday",
                  "day" : "2019-08-05",
                  "closes_at" : "2019-08-05T21:00:00.000-07:00",
                  "type" : "Summer quarter"
               },
               {
                  "day" : "2019-08-06",
                  "type" : "Summer quarter",
                  "closes_at" : "2019-08-06T21:00:00.000-07:00",
                  "closed" : false,
                  "opens_at" : "2019-08-06T08:00:00.000-07:00",
                  "weekday" : "Tuesday"
               },
               {
                  "closed" : false,
                  "weekday" : "Wednesday",
                  "opens_at" : "2019-08-07T08:00:00.000-07:00",
                  "day" : "2019-08-07",
                  "type" : "Summer quarter",
                  "closes_at" : "2019-08-07T21:00:00.000-07:00"
               },
               {
                  "closed" : false,
                  "opens_at" : "2019-08-08T08:00:00.000-07:00",
                  "weekday" : "Thursday",
                  "day" : "2019-08-08",
                  "type" : "Summer quarter",
                  "closes_at" : "2019-08-08T21:00:00.000-07:00"
               },
               {
                  "day" : "2019-08-09",
                  "type" : "Summer quarter",
                  "closes_at" : "2019-08-09T18:00:00.000-07:00",
                  "closed" : false,
                  "opens_at" : "2019-08-09T08:00:00.000-07:00",
                  "weekday" : "Friday"
               },
               {
                  "closes_at" : "2019-08-10T18:00:00.000-07:00",
                  "type" : "Summer quarter",
                  "notes" : "End Quarter Period",
                  "day" : "2019-08-10",
                  "opens_at" : "2019-08-10T10:00:00.000-07:00",
                  "weekday" : "Saturday",
                  "closed" : false
               }
            ]
         },
         "links" : {
            "self" : "http://127.0.0.1:3000/libraries/green/locations/media-microtext-center"
         },
         "type" : "locations"
      },
      {
         "id" : "green/privileges-desk",
         "relationships" : {
            "library" : {
               "links" : {
                  "related" : "http://127.0.0.1:3000/libraries/green"
               },
               "data" : {
                  "type" : "libraries",
                  "id" : "green"
               }
            }
         },
         "attributes" : {
            "name" : "Privileges desk",
            "hours" : [
               {
                  "closes_at" : "2019-08-04T20:00:00.000-07:00",
                  "type" : "Summer quarter",
                  "day" : "2019-08-04",
                  "opens_at" : "2019-08-04T12:00:00.000-07:00",
                  "weekday" : "Sunday",
                  "closed" : false
               },
               {
                  "type" : "Summer quarter",
                  "closes_at" : "2019-08-05T21:00:00.000-07:00",
                  "day" : "2019-08-05",
                  "opens_at" : "2019-08-05T08:00:00.000-07:00",
                  "weekday" : "Monday",
                  "closed" : false
               },
               {
                  "day" : "2019-08-06",
                  "type" : "Summer quarter",
                  "closes_at" : "2019-08-06T21:00:00.000-07:00",
                  "closed" : false,
                  "weekday" : "Tuesday",
                  "opens_at" : "2019-08-06T08:00:00.000-07:00"
               },
               {
                  "closed" : false,
                  "weekday" : "Wednesday",
                  "opens_at" : "2019-08-07T08:00:00.000-07:00",
                  "day" : "2019-08-07",
                  "closes_at" : "2019-08-07T21:00:00.000-07:00",
                  "type" : "Summer quarter"
               },
               {
                  "closes_at" : "2019-08-08T21:00:00.000-07:00",
                  "type" : "Summer quarter",
                  "day" : "2019-08-08",
                  "opens_at" : "2019-08-08T08:00:00.000-07:00",
                  "weekday" : "Thursday",
                  "closed" : false
               },
               {
                  "day" : "2019-08-09",
                  "closes_at" : "2019-08-09T18:00:00.000-07:00",
                  "type" : "Summer quarter",
                  "closed" : false,
                  "opens_at" : "2019-08-09T08:00:00.000-07:00",
                  "weekday" : "Friday"
               },
               {
                  "opens_at" : "2019-08-10T10:00:00.000-07:00",
                  "weekday" : "Saturday",
                  "closed" : false,
                  "closes_at" : "2019-08-10T18:00:00.000-07:00",
                  "type" : "Summer quarter",
                  "notes" : "End Quarter Period",
                  "day" : "2019-08-10"
               }
            ]
         },
         "type" : "locations",
         "links" : {
            "self" : "http://127.0.0.1:3000/libraries/green/locations/privileges-desk"
         }
      },
      {
         "attributes" : {
            "hours" : [
               {
                  "day" : "2019-08-04",
                  "closed" : true,
                  "weekday" : "Sunday"
               },
               {
                  "day" : "2019-08-05",
                  "weekday" : "Monday",
                  "closed" : true
               },
               {
                  "closed" : true,
                  "weekday" : "Tuesday",
                  "day" : "2019-08-06"
               },
               {
                  "closed" : true,
                  "weekday" : "Wednesday",
                  "day" : "2019-08-07"
               },
               {
                  "day" : "2019-08-08",
                  "closed" : true,
                  "weekday" : "Thursday"
               },
               {
                  "day" : "2019-08-09",
                  "closed" : true,
                  "weekday" : "Friday"
               },
               {
                  "day" : "2019-08-10",
                  "weekday" : "Saturday",
                  "closed" : true
               }
            ],
            "name" : "SSDS walk-in consulting"
         },
         "type" : "locations",
         "links" : {
            "self" : "http://127.0.0.1:3000/libraries/green/locations/ssds-walk-in-consulting"
         },
         "id" : "green/ssds-walk-in-consulting",
         "relationships" : {
            "library" : {
               "data" : {
                  "type" : "libraries",
                  "id" : "green"
               },
               "links" : {
                  "related" : "http://127.0.0.1:3000/libraries/green"
               }
            }
         }
      },
      {
         "attributes" : {
            "hours" : [
               {
                  "type" : "Summer quarter",
                  "day" : "2019-08-04",
                  "weekday" : "Sunday",
                  "closed" : true
               },
               {
                  "type" : "Summer quarter",
                  "closes_at" : "2019-08-05T17:00:00.000-07:00",
                  "day" : "2019-08-05",
                  "weekday" : "Monday",
                  "opens_at" : "2019-08-05T08:00:00.000-07:00",
                  "closed" : false
               },
               {
                  "day" : "2019-08-06",
                  "type" : "Summer quarter",
                  "closes_at" : "2019-08-06T17:00:00.000-07:00",
                  "closed" : false,
                  "opens_at" : "2019-08-06T08:00:00.000-07:00",
                  "weekday" : "Tuesday"
               },
               {
                  "day" : "2019-08-07",
                  "closes_at" : "2019-08-07T17:00:00.000-07:00",
                  "type" : "Summer quarter",
                  "closed" : false,
                  "weekday" : "Wednesday",
                  "opens_at" : "2019-08-07T08:00:00.000-07:00"
               },
               {
                  "weekday" : "Thursday",
                  "opens_at" : "2019-08-08T08:00:00.000-07:00",
                  "closed" : false,
                  "closes_at" : "2019-08-08T17:00:00.000-07:00",
                  "type" : "Summer quarter",
                  "day" : "2019-08-08"
               },
               {
                  "day" : "2019-08-09",
                  "closes_at" : "2019-08-09T17:00:00.000-07:00",
                  "type" : "Summer quarter",
                  "closed" : false,
                  "weekday" : "Friday",
                  "opens_at" : "2019-08-09T08:00:00.000-07:00"
               },
               {
                  "day" : "2019-08-10",
                  "notes" : "End Quarter Period",
                  "type" : "Summer quarter",
                  "closed" : true,
                  "weekday" : "Saturday"
               }
            ],
            "name" : "Velma Denning Room"
         },
         "type" : "locations",
         "links" : {
            "self" : "http://127.0.0.1:3000/libraries/green/locations/velma-denning-room"
         },
         "id" : "green/velma-denning-room",
         "relationships" : {
            "library" : {
               "links" : {
                  "related" : "http://127.0.0.1:3000/libraries/green"
               },
               "data" : {
                  "type" : "libraries",
                  "id" : "green"
               }
            }
         }
      }
   ]
}
```

🤷‍♂

### CODE REVIEW

**Comment by mejackreed** [comment]
I'm 👍  on adding this.

File: app/views/libraries/_library.json.jbuilder [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+    json.data do
+      json.array!(library.locations.with_hours) do |location|
+        json.type 'locations'
+        json.id "#{library.slug}/#{location.slug}"
+      end
+    end
+
+    json.links do
+      json.related library_locations_url(library)
```
Comment (mejackreed) [inline app/views/libraries/_library.json.jbuilder]: I can't seem to find any guidance from JSONAPI, but I assume that its understood that any related links will need to be requested with a JSON format or request header?

**Comment by mejackreed** [comment]
I realize not necessarily part of this PR, but should we put some limits on `from` / `to`?

File: app/views/libraries/_library.json.jbuilder (hunk 1 above)
Comment (cbeer) [reply app/views/libraries/_library.json.jbuilder]: I assume so too:

> JSON:API requires use of the JSON:API media type for exchanging data.

**Comment by cbeer** [comment]
=> #170

**Comment by jkeck** [comment]
I'm 👍 to this.

**Comment by camillevilla** [comment]
This looks good to me! It would be nice to send this to the folks at med and law for their feedback.

**Pull request merged**